src/encoded/static/components/browse/components/CustomColumnController.js

```diff
@@ -21,33 +23,54 @@ export class CustomColumnController extends React.Component {
         this.addHiddenColumn = this.addHiddenColumn.bind(this);
         this.removeHiddenColumn = this.removeHiddenColumn.bind(this);
         this.state = {
-            'hiddenColumns' : Array.isArray(props.defaultHiddenColumns) ? props.defaultHiddenColumns : []
+            'hiddenColumns' : _.clone(props.defaultHiddenColumns || {})
         };
     }
-    /*
+
     componentWillReceiveProps(nextProps){
         if (nextProps.defaultHiddenColumns !== this.props.defaultHiddenColumns){
-            this.setState({ hiddenColumns : _.uniq(this.state.hiddenColumns.concat(this.props.defaultHiddenColumns)) });
+            this.setState({ 'hiddenColumns' : _.clone(nextProps.defaultHiddenColumns || {}) });
         }
     }
-    */
+
+    /**
+     * @param {{ hiddenColumns?: string[], defaultHiddenColumns }} props - Component props.
+     * @returns {Object.<boolean>} Map of field names to boolean representing hidden or not.
+     */
     getAllHiddenColumns(props = this.props){
-        if (Array.isArray(props.hiddenColumns)) return props.hiddenColumns.concat(this.state.hiddenColumns);
+        if (Array.isArray(props.hiddenColumns)){
+            return _.extend(_.object(_.map(props.hiddenColumns, function(field){
+                return [ field, true ];
+            })), this.state.hiddenColumns);
+        }
         else return this.state.hiddenColumns;
     }
 
-    addHiddenColumn(fieldKey){
-        var hiddenCols = this.state.hiddenColumns.slice(0);
-        hiddenCols.push(fieldKey);
-        this.setState({ 'hiddenColumns' : hiddenCols });
+    addHiddenColumn(field){
+        this.setState(function(currState){
+            if (currState.hiddenColumns[field] === true){
+                return null;
+            }
+            var hiddenColumns = _.clone(currState.hiddenColumns);
```

Comment:
Why do you clone objects before editing them? To avoid concurrency/race conditions?

Reply:
In React and functional programming paradigms in general <sup>(1)</sup> it is bad practice (in terms of maintainability / reasoning about the code/system) to mutate the input that is passed in. This is very important with React especially because of the component tree structure -- an input variable may be passed in as a prop to multiple different components and if any one of those components mutates that input variable <sup>(2)</sup> -- without performing an appropriate state update or similar to inform other components of that change -- then will often start to receive unexpected results.

Examples of unexpected results include case where 2+ separate child components might transform the same input variable/prop differently and then pass it down to other components or elements which require different transformed structures -- at least one of those components which requires transformed structure would likely throw an exception. Smaller but harder to pinpoint issues will likely occur such as parent component not updating (because no state update triggered) in response to the data structure being transformed further down the rendering tree by a child component until something else triggers parent component update/rerender which could cause things to unexpectedly change on page (or errors) as parent component attempts to rerender something based off of the transformed-by-child variable (if said variable is a prop of or held in parent component state). And so on.

In our case in particular this is important because the `columns` are originally passed in from `context.columns` of the API endpoint response JSON, before this object of objects (_= object references_ **to same objects in context.columns -- unless they're cloned**) references gets converted into an array of object (references):
![](https://i.gyazo.com/fe69f19a281c1f12f5402f28ed375be6.png)

This `context` structure is rendered out in the DOM via the `BodyElement` Component:
![](https://i.gyazo.com/f6ae9d6b4d2fb440927a0d6bc8ae6747.png)

If the `context` (or any of its properties, via reference) get modified in place, then there is a chance that we will start to get errors in between server-side rendering and client-side rendering as that string will change - perhaps unexpectedly - and differ from the response JSON that was actually output by API endpoint. 


<sup>(1)</sup> React is largely a functional programming paradigm in that components take in input (props) and return output (JSX of HTML to be rendered in DOM).
<sup>(2)</sup> Applicable data structures which can be mutated include: object, array, object/array referenced from object property, object/array as an array element, etc.

Reply:
I didn't think about the situation with 2 child components editing the same object, but that does make sense. Thanks for the explanation.